proposals/user-scripts-api.md

```diff
@@ -0,0 +1,189 @@
+  - Scripts registered via the content_scripts key in the manifest file
+  - Scripts registered via [`scripting.registerContentScripts()`](https://developer.chrome.com/docs/extensions/reference/scripting/#method-registerContentScripts), following the order they were registered in. Updating a content script doesn't change its registration order.
+  - Scripts registered via `userScripts.register()`, following the order they were registered in. Updating a user script doesn’t change its registration order.
+- User scripts are always persisted across sessions, since the opposite behavior would be uncommon. (We may explore providing an option to customize this in the future.)
+
+### Browser level restrictions
+
+From here, each browser vendor should be able to implement their own restrictions.  Chrome is exploring limiting the access to this API when the user has enabled developer mode (bug), but permission grants are outside of the scope of this API proposal.
+
+## (Potential) Future Enhancements
+
+### `USER_SCRIPT`/ `ISOLATED` World Communication
+
+In the future, we may want to provide a more straightforward path for communication with a `USER_SCRIPT` world (as opposed to the [`window.postMessage()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) approach highlighted above).
+
+### Separate `USER_SCRIPT` Worlds
+
+In addition to specifying the execution world of `USER_SCRIPT`, we could allow extensions to inject in unique worlds by providing an identifier.  Scripts injected with the same identifier would inject in the same world, while scripts with different world identifiers inject in different worlds.  This would allow for greater isolation between user scripts (if, for instance, the user had multiple unrelated user scripts injecting on the same page).
+
+### Execute user scripts one time
+
+Currently, user scripts are registered and executed every time it matches the origin in a persistent way.  We may explore a way to execute a user script only one time to provide a new capability to user scripts (e.g `browser.userScripts.execute()`).
+
+### Establish common behaviors for the CSP of scripts injected into the main world by an extension
+
+Create certain HTML elements even if their src, href or contents violates CSP of the page so that the users don't have to nuke the site's CSP header altogether.
+
+### Dedicated API to execute code in the `MAIN` world from another world
+
+Provide an explicit API for an extension to execute code on the main world from a `USER_SCRIPT` or `ISOLATED` world.
+
+### Direct access to the `MAIN` world from another world
+
+Provide a way for the `USER_SCRIPT` or `ISOLATED` world to directly access the MAIN world (e.g. via an iframe contentWindow-like object).
+
+## Discussion Guidelines
+
+Please keep discussion to the first iteration introduced by this proposal.  Future enhancements can be tracked in the WECG [issue tracker](https://github.com/w3c/webextensions/issues).
```

Comment:
```suggestion
This is the first iteration of the userScripts API designed in https://github.com/w3c/webextensions/issues/279.
Future enhancements can be tracked in the WECG [issue tracker](https://github.com/w3c/webextensions/issues).
```

After this PR is merged, the discussion suggestion looks strange. Here is a suggested rephrase, but feel free to edit further or drop the section altogether.